File: problems-adapt/0139_vocabulary-segmentation/solution_dp.cpp
class Solution {
  public:
    bool canSegment(string s, vector<string> &vocabulary) {
        unordered_set<string> words(vocabulary.begin(), vocabulary.end());
        int n = (int)s.size();
        // reachable[i]: the prefix s[0..i) can be segmented; the empty prefix
        // is trivially segmentable.
        vector<bool> reachable(n + 1, false);
        reachable[0] = true;
        for (int i = 1; i <= n; i++) {
            // Any segmentation of s[0..i) ends with a last word s[j..i).
            for (int j = 0; j < i; j++) {
                if (reachable[j] && words.find(s.substr(j, i - j)) != words.end()) {
                    reachable[i] = true;
                    // Only feasibility matters, so stop at the first split.
                    break;
                }
            }
        }
        return reachable[n];
    }
};

File: problems/0101-0200/0139_word-break/solution_dp.py
from typing import List, Optional


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        words = set(wordDict)
        n = len(s)
        # reachable[i]: the prefix s[0..i) can be segmented.
        reachable = [False] * (n + 1)
        # The empty prefix is trivially segmentable.
        reachable[0] = True
        for i in range(1, n + 1):
            # Any segmentation of s[0..i) ends with a last word s[j..i).
            for j in range(i):
                if reachable[j] and s[j:i] in words:
                    reachable[i] = True
                    # Only feasibility matters, so stop at the first split.
                    break
        return reachable[n]
